<header class="qworld">
	<a
		class="qworld__logo"
		href="https://qworld.net/"
		target="_blank"
		rel="noopener noreferrer"
		aria-label="QWorld website"
	>
		<img src="../../assets/qworld.png" alt="QWorld logo">
	</a>
	<nav class="qworld__social" aria-label="QWorld social links">
		<ul>
			<li><a href="https://discord.gg/akCvr7U87g" target="_blank" rel="noopener noreferrer"><img src="../../assets/social-icons/discord.svg" alt="Discord icon"></a></li>
			<li><a href="https://www.youtube.com/qworld19" target="_blank" rel="noopener noreferrer"><img src="../../assets/social-icons/youtube.svg" alt="YouTube icon"></a></li>
			<li><a href="https://www.facebook.com/qworld19/" target="_blank" rel="noopener noreferrer"><img src="../../assets/social-icons/facebook.svg" alt="Facebook icon"></a></li>
			<li><a href="https://www.linkedin.com/company/qworld19/" target="_blank" rel="noopener noreferrer"><img src="../../assets/social-icons/linkedin.svg" alt="LinkedIn icon"></a></li>
		</ul>
	</nav>
</header>

$ \newcommand{\bra}[1]{\langle #1|} $
$ \newcommand{\ket}[1]{|#1\rangle} $
$ \newcommand{\braket}[2]{\langle #1|#2\rangle} $
$ \newcommand{\dot}[2]{ #1 \cdot #2} $
$ \newcommand{\biginner}[2]{\left\langle #1,#2\right\rangle} $
$ \newcommand{\mymatrix}[2]{\left( \begin{array}{#1} #2\end{array} \right)} $
$ \newcommand{\myvector}[1]{\mymatrix{c}{#1}} $
$ \newcommand{\myrvector}[1]{\mymatrix{r}{#1}} $
$ \newcommand{\mypar}[1]{\left( #1 \right)} $
$ \newcommand{\mybigpar}[1]{ \Big( #1 \Big)} $
$ \newcommand{\sqrttwo}{\frac{1}{\sqrt{2}}} $
$ \newcommand{\dsqrttwo}{\dfrac{1}{\sqrt{2}}} $
$ \newcommand{\onehalf}{\frac{1}{2}} $
$ \newcommand{\donehalf}{\dfrac{1}{2}} $
$ \newcommand{\hadamard}{ \mymatrix{rr}{ \sqrttwo & \sqrttwo \\ \sqrttwo & -\sqrttwo }} $
$ \newcommand{\vzero}{\myvector{1\\0}} $
$ \newcommand{\vone}{\myvector{0\\1}} $
$ \newcommand{\stateplus}{\myvector{ \sqrttwo \\  \sqrttwo } } $
$ \newcommand{\stateminus}{ \myrvector{ \sqrttwo \\ -\sqrttwo } } $
$ \newcommand{\myarray}[2]{ \begin{array}{#1}#2\end{array}} $
$ \newcommand{\X}{ \mymatrix{cc}{0 & 1 \\ 1 & 0}  } $
$ \newcommand{\I}{ \mymatrix{rr}{1 & 0 \\ 0 & 1}  } $
$ \newcommand{\Z}{ \mymatrix{rr}{1 & 0 \\ 0 & -1}  } $
$ \newcommand{\Htwo}{ \mymatrix{rrrr}{ \frac{1}{2} & \frac{1}{2} & \frac{1}{2} & \frac{1}{2} \\ \frac{1}{2} & -\frac{1}{2} & \frac{1}{2} & -\frac{1}{2} \\ \frac{1}{2} & \frac{1}{2} & -\frac{1}{2} & -\frac{1}{2} \\ \frac{1}{2} & -\frac{1}{2} & -\frac{1}{2} & \frac{1}{2} } } $
$ \newcommand{\CNOT}{ \mymatrix{cccc}{1 & 0 & 0 & 0 \\ 0 & 1 & 0 & 0 \\ 0 & 0 & 0 & 1 \\ 0 & 0 & 1 & 0} } $
$ \newcommand{\norm}[1]{ \left\lVert #1 \right\rVert } $
$ \newcommand{\pstate}[1]{ \lceil \mspace{-1mu} #1 \mspace{-1.5mu} \rfloor } $
$ \newcommand{\greenbit}[1] {\mathbf{{\color{green}#1}}} $
$ \newcommand{\bluebit}[1] {\mathbf{{\color{blue}#1}}} $
$ \newcommand{\redbit}[1] {\mathbf{{\color{red}#1}}} $
$ \newcommand{\brownbit}[1] {\mathbf{{\color{brown}#1}}} $
$ \newcommand{\blackbit}[1] {\mathbf{{\color{black}#1}}} $

<font style="font-size:28px;" align="left"><b>Two Probabilistic Bits  </b></font>
<br>
_prepared by Abuzer Yakaryilmaz_
<br><br>
[<img src="../qworld/images/watch_lecture.jpg" align="left">](https://youtu.be/ulbd-1c71sk)
<br><br><br>

Suppose that we have two probabilistic bits, and our probabilistic states respectively are

$ \begin{pmatrix} 0.2 \\ 0.8 \end{pmatrix} \quad{ and } \begin{pmatrix} 0.6 \\ 0.4 \end{pmatrix}. $

If we combine both bits as a single system, then what is the state of the combined system?

In total, we have four different states. We can name them as follows:
<ul>
    <li>00: both bits are in states 0</li>
    <li>01: the first bit is in state 0 and the second bit is in state 1</li>
    <li>10: the first bit is in state 1 and the second bit is in state 0</li>
    <li>11: both bits are in states 1</li>
</ul>

<h3> Task 1 </h3>

<b>Discussion and analysis:</b>

What are the probabilities of being in states $ 00 $, $ 01 $, $ 10 $, and $11$?

How can we represent these probabilities as a column vector?

<h3> Representation for states 0 and 1</h3>

The vector representation of state 0 is $ \begin{pmatrix} 1 \\ 0 \end{pmatrix}$. Similarly, the vector representation of state 1 is $ \begin{pmatrix} 0 \\ 1 \end{pmatrix} $.

We use $\lceil 0 \rfloor$  to represent $ \begin{pmatrix} 1 \\ 0 \end{pmatrix} $ and $\lceil 1 \rfloor$   to represent $\begin{pmatrix} 0 \\ 1 \end{pmatrix}$.

Then, the probabilistic state $ \begin{pmatrix} 0.2 \\ 0.8 \end{pmatrix} $ is also represented as $ 0.2 \lceil 0 \rfloor + 0.8 \lceil 1 \rfloor $.

Similarly, the probabilistic state $ \begin{pmatrix} 0.6 \\ 0.4 \end{pmatrix} $ is also represented as $ 0.6 \lceil 0 \rfloor + 0.4 \lceil 1 \rfloor $.

<h3> Composite systems </h3>

When two systems are composed, then their states are tensored to calculate the state of composite system.

The probabilistic state of the first bit is $  \begin{pmatrix} 0.2 \\ 0.8 \end{pmatrix} = $ $0.2 \lceil 0 \rfloor + 0.8 \lceil 1 \rfloor $.

The probabilistic state of the second bit is $  \begin{pmatrix} 0.6 \\ 0.4 \end{pmatrix} = 0.6 \lceil 0 \rfloor + 0.4 \lceil 1 \rfloor $.

Then, the probabilistic state of the composite system is $ \big( 0.2 \lceil 0 \rfloor + 0.8 \lceil 1 \rfloor \big) \otimes  \big( 0.6 \lceil 0 \rfloor + 0.4 \lceil 1 \rfloor \big) $.


<h3> Task 2 </h3>

Find the probabilistic state of the composite system.

<i> 
Rule 1: Tensor product distributes over addition in the same way as the distribution of multiplication over addition.

Rule 2: $ \big( 0.3 \lceil 1 \rfloor \big) \otimes \big( 0.7 \lceil 0 \rfloor \big) = (0.3 \cdot 0.7) \big( \lceil 1 \rfloor \otimes \lceil 0 \rfloor \big) = 0.21 \lceil 10 \rfloor $.
</i>

In [ ]:
0.12|00> + 0.08|01> + 0.48|10> + 0.32|11>

<a href="CS24_Two_Probabilistic_Bits_Solutions.ipynb#task2">click for our solution</a>

<h3> Task 3</h3>

Find the probabilistic state of the composite system by calculating this tensor product $ \begin{pmatrix} 0.2 \\ 0.8 \end{pmatrix} \otimes \begin{pmatrix} 0.6 \\ 0.4 \end{pmatrix} $.

In [ ]:
[0.12 0.08 0.48 0.32]

<a href="CS24_Two_Probabilistic_Bits_Solutions.ipynb#task3">click for our solution</a>

<h3> Task 4</h3>

Find the vector representations of $ \lceil 00 \rfloor $, $ \lceil 01 \rfloor $, $\lceil 10 \rfloor$, and $ \lceil 11 \rfloor $.

<i>The vector representation of $ \lceil ab \rfloor $ is $ \lceil a \rfloor \otimes \lceil b \rfloor $ for $ a,b \in \{0,1\} $.</i>

<a href="CS24_Two_Probabilistic_Bits_Solutions.ipynb#task4">click for our solution</a>

In [ ]:
|00> = [1 0 0 0]
|01> = [0 1 0 0]
|10> = [0 0 1 0]
|11> = [0 0 0 1]

---

<h3> Extra: Task 5 </h3>

Suppose that we have three bits.

Find the vector representations of $ \lceil abc \rfloor $ for each $ a,b,c \in \{0,1\} $.

In [ ]:
[a1b1c1, a1b1c2, a1b2c1, a1b2c2, a2b1c1, a2b1c2, a2b2c1, a2b2c2]

In [1]:
def bit_vector(bit):
	return [1, 0] if bit == 0 else [0, 1]

def tensor(v1, v2):
	return [x * y for x in v1 for y in v2]

for a in (0, 1):
	for b in (0, 1):
		for c in (0, 1):
			vec = tensor(tensor(bit_vector(a), bit_vector(b)), bit_vector(c))
			print(f"{a}{b}{c}:", vec)

000: [1, 0, 0, 0, 0, 0, 0, 0]
001: [0, 1, 0, 0, 0, 0, 0, 0]
010: [0, 0, 1, 0, 0, 0, 0, 0]
011: [0, 0, 0, 1, 0, 0, 0, 0]
100: [0, 0, 0, 0, 1, 0, 0, 0]
101: [0, 0, 0, 0, 0, 1, 0, 0]
110: [0, 0, 0, 0, 0, 0, 1, 0]
111: [0, 0, 0, 0, 0, 0, 0, 1]


<h3> Extra: Task 6 </h3>

<i>This task is challenging.</i>

Suppose that we have four bits. 

Number 9 is represented as $ 1001 $ in binary. Verify that the vector representation of $ \lceil 1001 \rfloor $ is the zero vector except its $10$th entry, which is 1.

Number 7 is represented as $ 0111 $ in binary. Verify that the vector representation of $ \lceil 0111 \rfloor $ is the zero vector except its $8$th entry, which is 1.

Generalize this idea for any number between 0 and 15.

Generalize this idea for any number of bits.